<a href="https://colab.research.google.com/github/Abdurrasyiid/Dasar-Visualisasi-Data/blob/master/Rata_Rata_Statistik.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Rata - Rata Statistik** 

# Rata - Rata Aritmatika

Mean atau sering kita kenal dengan nilai rata-rata mengacu pada nilai yang dalam statistik kita sebut dengan rata-rata aritmatika atau arithmetic mean. Masih ingat cara menghitung rata-rata? Dalam menghitung rata-rata kita mulai dengan menjumlahkan semua elemen data yang ada. Selanjutnya hasil penjumlahan tersebut kita bagi dengan total elemen yang ada. Misalnya kita mempunyai data sebagai berikut:

Pada sebuah ulangan matematika, nilai yang didapatkan oleh 10 siswa sebagai berikut:



```
65,70,75,70,80,60,80,90,70,65
```

Berapa rata-rata (arithmetic mean) nilai ulangan matematika dari 10 siswa tersebut?

Apabila kita menggunakan cara perhitungan seperti yang sudah dijelaskan sebelumnya, maka rumusnya seperti berikut:

> $x=x_{1} + x_{2} + x_{3} + .... + x_{n} = \frac{1}{n} \sum_{i=1}^{n} x_{i}$

Sehingga,

>$\frac{65 + 70 + 75 + 70 + 80 + 60 + 80 + 90 + 70 + 65}{10}=72,5$

In [1]:
x = [65,70,75,70,80,60,80,90,70,65]
y = len(x)
jum = 0

for i in x :
  jum+=i

Mean = jum/y
print(format(Mean,".2f"))

72.50


# Rata - rata Harmonik

Rata-rata harmonik bisa dikatakan sebagai kebalikan dari rata-rata aritmatika. Kenapa demikian? Karena jumlah data dibagi dengan pecahan data ke-i dengan pembilang 1 (1/xi). Apabila dituliskan dalam rumus matematika menjadi sebagai berikut:

$H=\frac{n}{\sum_{i=1}^{n} (\frac{1}{x_i})}$

Diketahui bahwa:

H = rata-rata harmonik

n = banyak data

$x_{i}$ = nilai data ke- ${i}$

Untuk memperjelas rumus di atas, berikut terdapat contoh beberapa angka:

```
2,3,5,6,6
```
Dari angka di atas apabila ingin didapatkan rata-rata harmoniknya maka penerapannya dengan rumus sebagai berikut:

$H=\frac{5} {\frac{1}{2}+\frac{1}{3}+\frac{1}{5}+\frac{1}{6}+\frac{1}{6}}$
$H = 3,66 $

---
KASUS 1

rata-rata harmonik memiliki kelebihan untuk beberapa kasus data misalnya data kecepatan dan data yang mengandung outlier.

Jika kita lihat referensi-referensi mengenai rata-rata harmonik, maka akan kita dapati contoh data yang digunakan dalam menghitung rata-rata harmonik adalah data kecepatan. Rata-rata kecepatan memang sebaiknya dihitung menggunakan rata-rata harmonik.

 

Kenapa demikian? Coba simak ilustrasi berikut.


---
> Misalkan Alesha menegendarai mobil dari kota A ke kota B yang berjarak 60 km dengan kecepatan 60 km/jam. Sesampai di kota B, Alesha langsung kembali lagi ke kota A dengan kecepatan 30 km/jam. Berapakah rata-rata kecepatan mobil yang dikendarai oleh Alesha dari kota A ke kota B dan kembali lagi ke kota A?



>Jika rata-rata yang kita gunakan adalah rata-rata aritmatik, maka nilai rata-rata kecepatan mobil yang dikendarai Alesha adalah $\frac{1}{2}(60+30)=45$ 

>Sebenarnya nilai rata-rata ini ***bias*** karena tidak menggambarkan rata-rata yang sebenarnya.
---

**Kenapa bisa bias?**

Begini penjelasannya.


---


Jika Alesha mengendarai mobil dari kota A ke kota B yang berjarak 60 km dengan kecepatan 60 km/jam, maka waktu tempuh yang ia butuhkan adalah 1 jam. Kemudian ia kembali lagi ke kota A dengan kecepatan 30 km/jam. Dengan kecepatan 30 km/jam tersebut, waktu yang dibutuhkan Alesha untuk kembali adalah 2 jam. Artinya jarak perjalanan yang ditempuh Alesha adalah 120 km (60 km pergi dan 60 km pulang) dan waktu tempuh yang dibutuhkannya adalah 3 jam (1 jam pergi dan 2 jam pulang). Dengan demikian kecepatan mobil Alesha pergi-pulang adalah $\frac{120\text{ km}}{3\text{ jam}}=40$ 

Kecepatan mobil yang dikendarai Alesha pergi-pulang ternyata berbeda dengan hasil hitungan rata-rata aritmatik. Oleh karena itu rata-rata aritmatik tidak tepat dalam mengestimasi rata-rata sebenarnya.

---

Bagaimana dengan rata-rata harmonik?

Mari kita lakukan penghitungan dengan menggunakan rata-rata harmonik.

\begin{aligned} H&=\frac{n}{\displaystyle\sum_{i=1}^n\frac{1}{x_i}}\\ &=\frac{2}{\displaystyle\frac{1}{60}+\frac{1}{30}}\\ &=40 \end{aligned} 

Nilai rata-rata harmonik ternyata sama dengan rata-rata sebenarnya yaitu 40 km/jam. Oleh karena itu untuk kasus seperti ini rata-rata harmonik lebih tepat digunakan daripada rata-rata aritmatik.

---
KASUS 2
--- ​Rata-rata Harmonik untuk Data Yang Mengandung Outlier (Pencilan)

---

Rata-rata harmonik sangat sensitif dengan data yang bernilai rendah. Kondisi ini membuat rata-rata harmonik sangat baik digunakan untuk data yang memiliki outlier dimana outlier tersebut memiliki nilai jauh lebih besar secara signifikan.

Misalkan diberikan data



```
4, 5, 4, 40, 3, 5, 6, 5
```



Rata-rata aritmatik dari data tersebut adalah 9. Rata-rata tersebut tentu kurang memberikan gambaran keterwakilan data. Jika kita gunakan rata-rata harmonik, maka rata-ratanya adalah 4,92. Rata-rata ini cukup memberikan gambaran keterwakilan data.

menghitung rata-rata harmonik pada data yang mengandung outlier di atas dapat disimpulkan bahwa rata-rata harmonik sangat baik digunakan jika sebagian besar nilai data terdistribusi secara merata tetapi terdapat beberapa outlier dengan nilai yang lebih besar secara signifikan.

In [2]:
#Kasus 1
kecepatan = [60,30]
n = len(kecepatan)
jumlah = 0
for i in kecepatan :
  jumlah+=1/i

Harmonik = n/jumlah
print(Harmonik)


40.0


In [3]:
#Kasus 2
x = [4, 5, 4, 40, 3, 5, 6, 5]
n = len(x)
jumlah = 0
for i in x :
  jumlah+=1/i

Harmonik = n/jumlah
print(format(Harmonik, ".2f"))

4.92


# Rata - Rata Ukur (Geometric Mean)

Rata-rata yang didapatkan dengan cara mengalikan semua data yang ada dalam suatu kelompok atau sampel. Kemudian data tersebut diakar pangkat dengan jumlah sampel yang ada. Karena menggunakan akar pangkat, apabila terdapat data yang memiliki nilai negatif maka perhitungan dengan rata-rata ukur (geometric mean) tidak bisa digunakan. Berikut rumus untuk menghitung rata-rata geometrik. 

***Metode pertama:***

$ G = (x_{1}. x_{2}.x_{n})^\frac{1}{n}$

Diketahui bahwa:

G = rata-rata geometrik

$x_{1}...x_{n}$ = data ke-i

n = banyak data

Metode pertama di atas adalah rumus yang bisa digunakan dengan mudah dan efektif apabila jumlah datanya tidak terlalu banyak.

Namun, jika datanya sangat banyak mencapai puluhan, ratusan, atau lebih pasti akan terasa sulit apabila menggunakan metode pertama tadi. Maka, terdapat cara kedua yaitu menggunakan logaritma. Hitung log semua data terlebih dulu, kemudian lakukan perhitungan aritmatika seperti biasanya. Lebih lengkapnya simak rumus berikut:

***Metode Kedua:***
Metode Logaritma

$Log (G) = \frac{1}{n} \sum_{i=1}^n log(x_{i})$

Diketahui bahwa:

Log (G) = rata-rata geometrik

$x_{i}$ = data ke-i

n = banyak data


Berikut contoh soal dari rata-rata ukur (geometric mean):

Hitunglah data di bawah ini dengan rata-rata ukur baik perhitungan biasa maupun logaritma.



```
3,4,6
```
Dihitung dengan perhitungan rata-rata ukur biasa seperti berikut:

$ G = (x_{1}. x_{2}.x_{n})^\frac{1}{n}$

$ G = (3. 4. 6)^\frac{1}{3}$

$G = 4.160$

Dihitung dengan perhitungan rata-rata ukur Logaritma seperti berikut:

$Log (G) = \frac{1}{n} \sum_{i=1}^n log(x_{i})$

$Log (G) = \frac{log(3)+log(4)+log(6)}{3}$

$ G = 10^{0.6191108321..}$

$ G = 4.160 $

Angka 10 pada 100.61911083214375616 merupakan default log yang berarti Log(10) = 1.

Sehingga dari data di atas dapat dilihat bahwa apabila datanya sederhana, lebih mudah dihitung dengan metode biasa yang pertama. Namun apabila datanya kompleks lebih baik menggunakan metode logaritma. Manapun pilihan Anda, hasilnya tetap sama.

In [4]:
#Metode pertama
import math
x = [3,4,6]
jum = 1
n = len(x)
y = 1/n
for i in x :
  jum*=i

z = math.pow(jum,y)
print(format(z,".3f"))

4.160


In [5]:
#Metode Kedua

x = [3,4,6]
n = len(x)
jum = 0
for i in x :
  jum+=math.log10(i)

z = jum/n
G = math.pow(10,z)
print(format(G,".3f"))

4.160


# Rata-rata tertimbang (Weighted Mean)

Pernah mendengar sebelumnya? Rata-rata tertimbang dihitung dengan memperhatikan bobot yang ada dalam setiap datanya. Setiap bobot tersebut merupakan pasangan setiap dari data. Hal ini sangat berguna ketika terdapat bobot tertentu pada suatu data yang bisa mempengaruhi data yang bersangkutan. Inilah perbedaan utama yang membedakan dengan perhitungan rata-rata aritmatika. Apabila dituliskan dalam rumus matematika sebagai berikut:

$x = \frac{\sum_{i=1}^n W_{i} . X_{i}}{\sum_{i=1}^n W_{i}} $

x  = rata-rata tertimbang (weighted mean)

xi = nilai data ke-i

wi = bobot data ke-i

n = jumlah data

Sederhananya seperti ini, terdapat sebuah sekolah yang menerima pendaftaran siswa baru. Salah satu penilaiannya adalah angka di rapor. Proses seleksi berfokus pada beberapa mata pelajaran tertentu dan memiliki bobot masing-masing dalam penilaiannya seperti, Matematika memiliki bobot 40, IPA bobot nilainya 30, Bahasa Indonesia bobot nilainya 15, dan Bahasa Inggris bobot nilainya 15. Untuk dapat diterima di sekolah tersebut, calon siswa setidaknya memiliki nilai raport rata-rata 70 berdasarkan bobot yang sudah ditentukan. 

Nah, terdapat dua orang calon siswa, sebut saja Budi dan Toni yang memiliki nilai raport sebagai berikut:

Budi:

>Matematika : 80

>IPA : 70

>Bahasa Indonesia : 70

>Bahasa Inggris : 60



Toni:

>Matematika : 60

>IPA : 70

>Bahasa Indonesia 80

>Bahasa Inggris 85



Di antara kedua siswa di atas, siapa yang kira-kira berpotensi untuk lolos seleksi?

Apabila dihitung dengan rata-rata tertimbang sebagai berikut:



---
Rata rata tertimbang Budi = $\frac{(40x80)+(30x75)+(15x70)+(15x60)}{40+30+15+15}$
= 74

---

---
Rata rata tertimbang Toni = $\frac{(40x60)+(30x70)+(15x80)+(15x85)}{40+30+15+15}$
= 69

---

**Bandingkan perhitungan rata - rata budi dan toni menggunakan Mean aritmatika**

---
Rata rata tertimbang Budi = $\frac{(80+75+70+60)}{4}$
= 71.5

---

---
Rata rata tertimbang Toni = $\frac{(60+70+80+85}{4}$
= 72.5

---

Jika menggunakan Rata - Rata Aritmatika maka kedua siswa tersebut lolos, sedangkan jika menggunakan Rata - Rata tertimbang siswa budi yang lolos sehingga disimpulkan bahwa untuk kasus ini perhitungan rata - rata menggunakan rata - rata tertimbang **cocok**, sedangkan rata - rata aritmatika **tidak cocok** 


In [14]:
#Rata - Rata Tertimbang

budi = [80,75,70,60]
lenb = len(budi)
Toni = [60,70,80,85]
lent = len(Toni)
bobot = [40,30,15,15]
jumb = 0 # sum(Wi x Xi)
jumt = 0 # sum(Wi x Xi)
jumbobot = 0 # sum Wi

# sum(Wi x Xi)
for i in range (0,lenb) :
  for j in range(i,i+1) :
    jumb+=(budi[i]*bobot[j])
    
# sum Wi
for i in bobot :
  jumbobot += i

# sum(Wi x Xi)
for i in range (0,lent) :
  for j in range(i,i+1) :
    jumt+=(Toni[i]*bobot[j])

nilaibudi=jumb/jumbobot
nilaitoni=jumt/jumbobot
print("Rata-Rata Budi :",nilaibudi)
print("Rata-Rata Toni :",nilaitoni)

if(nilaibudi >=70 and nilaitoni >=70) :
  print("keduanya lolos")
elif(nilaibudi>=70 and nilaitoni<70) :
  print("hanya budi yang lolos")
elif(nilaibudi<70 and nilaitoni>=70) :
  print("hanya Toni yang lolos")
else :
  print("tidak ada yang lolos")


Rata-Rata Budi : 74.0
Rata-Rata Toni : 69.75
hanya budi yang lolos


# Rata-rata terpotong (Trimmed Mean)

Rata-rata terpotong bisa diartikan dengan memotong atau memangkas data berdasarkan posisi data dengan nilai rata-rata. Misal, rata-rata terpotong sebanyak 5% berarti terdapat 5% nilai angka (setelah diurutkan dari yang terkecil ke terbesar) yang berada pada ujung kiri data (awal) dan ujung kanan data (akhir) yang harus dibuang. Dengan membuang nilai tersebut diharapkan dapat mengurangi outliers sehingga data yang didapatkan lebih rapi. Apa itu outliers? Dalam ilmu statistik, outlier atau sering disebut “pencilan” merupakan sebuah nilai angka yang berbeda atau menyimpang sangat jauh dari angka lainnya di dalam satu kelompok data. Tujuan dari perhitungan rata-rata terpotong digunakan supaya data yang dihasilkan lebih akurat dan konsisten.

Apabila ditulis dengan rumus sebagai berikut:

$ u = \frac{\sum X_{i}}{n}$

Diketahui dari rumus di atas bahwa:

Rata - Rata Terpotong = $\frac{Hasil penjumlahan data yang sudah terpotong}{Banyaknya data yang sudah terpotong}$

Misal terdapat data sebagai berikut:



```
3,5,7,8,10,15,20,25,40,70
```
Berapa rata-rata terpotong sebanyak 30% dari data di atas?

Sebelum memasukkan data tersebut ke dalam rumus, kita cari dulu data yang harus dihapus atau dipotong. Ibaratkan data yang terpotong dengan g, sehingga perhitungannya sebagai berikut.

```
g = (persen rata rata terpotong) x (banyaknya sample data)

g = 30% x 10

g = 3
```
Nah, dari perhitungan di atas kita mendapatkan nilai 3 yang berarti kita menghapus 3 data yang paling kanan dan 3 data yang paling kiri.

```
Data sebelum dipotong = 3,5,7,8,10,15,20,25,40,70

Data yang dipotong sebanyak 30% = (3,5,7)<-hapus  8,10,15,20, ->hapus(25,40,70)

Data yang dihasilkan = 8,10,15,20
```
$ u = \frac{\sum X_{i}}{n}$


Rata - Rata Terpotong = $\frac{8+10+15+20}{4}$ = 13,25




In [25]:
x = [40,3,7,20,10,70,8,25,5,15]
print("Data sebelum diurutkan :", x)
#Langkah pertama urutkan ascending list terlebih dahulu
x = sorted(x)
print("Data setelah diurutkan dan sebelum dipotong :", x)
#Langkah kedua masukkan variabel yg dibutuhkan
n = len(x)
g = 30/100 * n

#potong list dengan nilai potong yang didapatkan dari sisi kiri dan kanan
#Misal nilai potong (g) adalah 3 Sehingga hapus list 0-2 dan 7-9

#del 7-9
sisi_kanan = int(n-g)
del(x[sisi_kanan:])
#del 0-2
sisi_kiri = int(g)
del(x[:sisi_kiri])

print("Data yang dihasilkan :", x)
n_baru=len(x)
R = 0
for i in x :
  R+=i

Rata2 = R/n_baru

print("Rata - Rata Terpotong :", Rata2)


Data sebelum diurutkan : [40, 3, 7, 20, 10, 70, 8, 25, 5, 15]
Data setelah diurutkan dan sebelum dipotong : [3, 5, 7, 8, 10, 15, 20, 25, 40, 70]
Data yang dihasilkan : [8, 10, 15, 20]
Rata - Rata Terpotong : 13.25
